src/java/org/apache/cassandra/service/StorageProxy.java

```diff
@@ -1395,11 +1477,36 @@ public static void mutateAtomically(List<Mutation> mutations,
             ReplicaPlan.ForWrite batchlogReplicaPlan = ReplicaPlans.forBatchlogWrite(ClusterMetadata.current(), batchConsistencyLevel == ConsistencyLevel.ANY);
             final TimeUUID batchUUID = nextTimeUUID();
             boolean wroteToBatchLog = false;
+            Epoch lastEpoch = null;
             while (true)
             {
+                if (lastEpoch != null)
```

Comment:
this logic is very much copy/paste, can we refactor into a reusable function?  the only difference is a on-timeout throw handler

```
awaitPending(() -> {
  doFallibleWriteWithMetricTracking(() -> {throw new WriteTimeoutException(WriteType.BATCH, consistencyLevel, 0, 0, "Timed out waiting for updated cluster metadata");}, consistencyLevel);
});
```